subsys/bluetooth/audio/aics_client.c

```diff
@@ -0,0 +1,889 @@
+		if (inst->cli.cb->set_manual_mode) {
+			inst->cli.cb->set_manual_mode(conn, inst, err);
+		}
+		break;
+	case BT_AICS_OPCODE_SET_AUTO:
+		if (inst->cli.cb->set_auto_mode) {
+			inst->cli.cb->set_auto_mode(conn, inst, err);
+		}
+		break;
+	default:
+		BT_DBG("Unknown opcode 0x%02x", inst->cli.cp_val.cp.opcode);
+		break;
+	}
+}
+
+static uint8_t internal_read_state_cb(struct bt_conn *conn, uint8_t err,
+				      struct bt_gatt_read_params *params,
+				      const void *data, uint16_t length)
+{
+	int cb_err = err;
+	struct bt_aics *inst = lookup_aics_by_handle(conn, params->single.handle);
+	struct bt_aics_state *state = (struct bt_aics_state *)data;
+
+	memset(params, 0, sizeof(*params));
+
+	if (!inst) {
+		BT_ERR("Instance not found");
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (err) {
+		BT_WARN("Volume state read failed: %d", err);
+		cb_err = BT_ATT_ERR_UNLIKELY;
+	} else if (data) {
+		if (length == sizeof(*state)) {
+			int write_err;
+
+			BT_DBG("Gain %d, mute %u, gain_mode %u, counter %u",
+			       state->gain, state->mute,
+			       state->gain_mode, state->change_counter);
+			inst->cli.change_counter = state->change_counter;
+
+			/* clear busy flag to reuse function */
+			inst->cli.busy = false;
+
+			if (inst->cli.cp_val.cp.opcode == BT_AICS_OPCODE_SET_GAIN) {
+				write_err = bt_aics_client_gain_set(conn, inst,
+								    inst->cli.cp_val.gain_setting);
+			} else {
+				write_err = aics_client_common_control(conn,
+								       inst->cli.cp_val.cp.opcode,
+								       inst);
+			}
+
+			if (write_err) {
+				cb_err = BT_ATT_ERR_UNLIKELY;
+			}
+		} else {
+			BT_DBG("Invalid length %u (expected %zu)", length, sizeof(*state));
+			cb_err = BT_ATT_ERR_UNLIKELY;
+		}
+	}
+
+	if (cb_err) {
+		inst->cli.busy = false;
+		aics_cp_notify_app(conn, inst, cb_err);
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+
+static void aics_client_write_aics_cp_cb(struct bt_conn *conn, uint8_t err,
+					 struct bt_gatt_write_params *params)
+{
+	int cb_err = err;
+	struct bt_aics *inst = lookup_aics_by_handle(conn, params->handle);
+
+	memset(params, 0, sizeof(*params));
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return;
+	}
+
+	BT_DBG("Inst %p: err: %d", inst, cb_err);
+
+	if (cb_err == BT_AICS_ERR_INVALID_COUNTER && inst->cli.state_handle) {
+		inst->cli.read_params.func = internal_read_state_cb;
+		inst->cli.read_params.handle_count = 1;
+		inst->cli.read_params.single.handle = inst->cli.state_handle;
+		inst->cli.read_params.single.offset = 0U;
+
+		cb_err = bt_gatt_read(conn, &inst->cli.read_params);
+
+		if (cb_err) {
+			BT_WARN("Could not read Volume state: %d", cb_err);
+		}
+	}
+
+	inst->cli.busy = false;
+
+	aics_cp_notify_app(conn, inst, cb_err);
+}
+
+static int aics_client_common_control(struct bt_conn *conn, uint8_t opcode, struct bt_aics *inst)
+{
+	int err;
+
+	if (!conn) {
+		return -ENOTCONN;
+	} else if (!inst->cli.control_handle) {
+		BT_DBG("Handle not set for opcode %u", opcode);
+		return -EINVAL;
```

Comment:
"Handle not set for control point"?

Reply:
This is on purpose, since this is a common cp write function it's useful to log which opcode failed.

Reply:
Agree.  But the handle is the same for all opcodes, right?

Reply:
(I am OK as is, I just thought it might confuse people to believe that different opcodes have different handles.)